File: CPP/24. Swap Nodes in Pairs.cpp
#include <iostream>
#include <vector>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }

        ListNode* next = head->next;
        ListNode* rest_left = next->next;
        next->next = head;
        ListNode* rest_head = swapPairs(rest_left);
        head->next = rest_head;

        return next;
    }
};

int main() {
	Solution sol;
	ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
	ListNode* result = sol.swapPairs(head);
	printListNode(result);
	return 0;
}

File: Python/24. Swap Nodes in Pairs.py
from typing import Optional
from ListNode import *

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        dummy = ListNode()
        curr = head
        prev = dummy
        
        while curr and curr.next:
            prev.next = curr.next
            curr.next = prev.next.next
            prev.next.next = curr
            
            prev = curr
            curr =  curr.next
            
        return dummy.next

if __name__ == '__main__':
    head = ListNode(1,ListNode(2,ListNode(3,ListNode(4))))
    result = Solution().swapPairs(head)
    print(result)